### PR TITLE
[krel] krel promote-images subcommand

### LINKED ISSUE #1445: [krel] Introduce krel subcommand for image promotion
#### What would you like to be added:

Right now we draft a separate [image promotion PR](https://github.com/kubernetes/k8s.io/pull/1071) after `mock release`, which adds another manual step in release cutting. To make the overall process less error-prone, we can add an additional `krel` subcommand to do the image promotion PR for us. The PR should be created on hold and will be lifted right before `nomock release`, which can be done by krel, too. After the image promotion PR got merged, we still have to wait for the [post-submit job](https://prow.k8s.io/?job=post-k8sio-cip) to complete. Krel could watch this job and run the `nomock stage` only if the promotion was successful.

/priority important-soon

### PR DESCRIPTION
**Pull request by puerco**
Signed-off-by: Adolfo García Veytia (Puerco) <adolfo.garcia@uservers.net>

/kind feature

#### What this PR does / why we need it:

This PR adds a new subcommand to krel: `promote-images`. 

When run, `krel promote` takes a tag and runs [cip-mm](https://github.com/kubernetes-sigs/k8s-container-image-promoter/tree/master/cmd/cip-mm) to modify the image promotion manifests.

After writing the new image into the manifest, krel uses a fork of kubernetes/k8s.io`. It will clone it, write the manifests in it, commit and push the changes. Finally it will create a Pull Request with the modified files.

The subcommand has an `--interactive` flag. When set, krel will ask the user to confirm before running every step. 

The command can be run and will not push or create a PR if no changes are made to the current manifest.

As part of this enhancement, a new API has been added to the release package to read and write the image list, iteself part of a thin manifest from the imagee promoter. Tests are included for all the new functions.

```
Usage:
  krel promote-images [flags]

Flags:
      --fork string   the user's fork of kubernetes/k8s.io
  -h, --help          help for promote-images
  -i, --interactive   interactive mode, asks before every step
  -t, --tag string    version tag of the images we will promote
```

#### Which issue(s) this PR fixes:
fixes #1445 

#### Special notes for your reviewer:

This PR depends on #1543 to run

It needs to have `cip-mm` in the path to run.

#### Does this PR introduce a user-facing change?
```release-note
- New `krel promote-images` subcommand to create the image promotion pull requests
- API to read and write image lists in an [image promoter thin-manifest](https://github.com/kubernetes-sigs/k8s-container-image-promoter#thin-manifests-example)
```

### CODE REVIEW

**Comment by justaugustus** [comment]
Awesome! I've got a few questions, but I'm going to hold review for a weekday.
/hold
/assign

File: cmd/krel/cmd/promote.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+	Short: "Starts an image promotion for a tag of kubernetes images",
+	Long: `krel promote
+
+The 'promote' subcommand of krel updates the image promoter manifests
+ans creates a PR in kubernetes/k8s.io`,
+	SilenceUsage:  true,
+	SilenceErrors: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		// Run the PR creation function
+		return runPromote()
```
Comment (saschagrunert) [inline cmd/krel/cmd/promote.go:53]: I suggest that we pass the `promoteOptions` in here to support testing.

File: cmd/krel/cmd/promote.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+		false,
+		"interactive mode, asks before every step",
+	)
+
+	if err := promoteCommand.MarkPersistentFlagRequired("tag"); err != nil {
+		logrus.Error(errors.Wrap(err, "marking tag flag as required"))
+	}
+	if err := promoteCommand.MarkPersistentFlagRequired("fork"); err != nil {
+		logrus.Error(errors.Wrap(err, "marking tag flag as required"))
+	}
```
Comment (saschagrunert) [inline cmd/krel/cmd/promote.go:122]: Let's refactor this into a loop :)

File: cmd/krel/cmd/promote.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+	if o.tag == "" {
+		return errors.New("cannot start promotion --tag is required")
+	}
+	if o.userFork == "" {
+		return errors.New("cannot start promotion --fork is required")
+	}
+
+	// Check the fork slug
+	_, _, err := git.ParseRepoSlug(o.userFork)
+	if err != nil {
```
Comment (saschagrunert) [inline cmd/krel/cmd/promote.go:73]:
```suggestion
	if _, _, err := git.ParseRepoSlug(o.userFork); err != nil {
```

File: cmd/krel/cmd/promote.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+		return errors.Wrap(err, "checking user's fork")
+	}
+
+	// Verify we got a valid tag
+	_, err = util.TagStringToSemver(o.tag)
+	if err != nil {
```
Comment (saschagrunert) [inline cmd/krel/cmd/promote.go:79]:
```suggestion
	if _, err := util.TagStringToSemver(o.tag); err != nil {
```

File: cmd/krel/cmd/promote.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+	rootCmd.AddCommand(promoteCommand)
+}
+
+func runPromote() error {
+	// Validate options
+	branchname := promoteOpts.tag + promotionBranchSuffix
+
+	// Check the cmd line opts
+	if err := promoteOpts.Validate(); err != nil {
+		return errors.Wrap(err, "")
```
Comment (saschagrunert) [inline cmd/krel/cmd/promote.go:133]: This should include a message I guess :)

**Review by saschagrunert — COMMENTED** [review]
Just found some nits. :+1: :blush:

**Comment by saschagrunert** [comment]
Theoretically, it should now succeed because the required PR got merged:
/retest

File: cmd/krel/cmd/promote.go (hunk 1 above)
Comment (puerco) [reply cmd/krel/cmd/promote.go]: Done :+1:

File: cmd/krel/cmd/promote.go (hunk 5 above)
Comment (puerco) [reply cmd/krel/cmd/promote.go]: Duh, yes I did not notice that one. Fixed.

**Comment by puerco** [comment]
I've renamed the command as `promote-images`. I've addressed all nits and tests are green now :)

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Pull request merged**